### PR TITLE
Refactor the Particles API

### PR DESCRIPTION
**Pull request by Cybermaxke**
API | [Common](https://github.com/SpongePowered/SpongeCommon/pull/702)

Refactor of the particles api to simplify the process to create particle because of the different types of particles (colorable, resizable), all the sub classes are removed and replaced by a option system which will also make it easier to share options between different particle types. Additionally added particle effect for 1.9 and renamed the particle fields.

Changes:
- Removed the particle subclasses (type, builders and effect subclasses)
- Added a option system for colors, scales, notes
- Fixed some names like: motion -> velocity, size -> scale
- Renamed particle types fields to have more readable names
- Added missing 1.9 particle types

Example plugin:
```java
@Plugin(id = "particletest")
public class ParticleTestPlugin {

    private static final String PERMISSION = "particletest.command";

    @Inject
    private Logger logger;

    @Listener
    public void onGameInit(GameInitializationEvent event) {
        this.logger.info("Loaded the particle test plugin!");
        Sponge.getCommandManager().register(this, CommandSpec.builder()
                .arguments(
                        GenericArguments.catalogedElement(Text.of("type"), ParticleType.class),
                        GenericArguments.vector3d(Text.of("position")),
                        GenericArguments.optional(GenericArguments.world(Text.of("world"))),
                        GenericArguments.flags()
                                .valueFlag(GenericArguments.integer(Text.of("amount")), "-amount", "a")
                                .valueFlag(GenericArguments.vector3d(Text.of("offset")), "-offset", "o")
                                .valueFlag(GenericArguments.vector3d(Text.of("motion")), "-motion", "m")
                                .valueFlag(GenericArguments.vector3d(Text.of("color")), "-color", "c")
                                .valueFlag(GenericArguments.doubleNum(Text.of("scale")), "-scale", "s")
                                .valueFlag(GenericArguments.catalogedElement(Text.of("note"), NotePitch.class), "-note", "n")
                                .valueFlag(GenericArguments.catalogedElement(Text.of("block"), BlockState.class), "-block", "b")
                                .valueFlag(GenericArguments.catalogedElement(Text.of("item"), ItemType.class), "-item", "i")
                                .buildWith(GenericArguments.none()))
                .permission(PERMISSION)
                .executor((src, args) -> {
                    ParticleType particleType = args.<ParticleType>getOne("type").get();
                    Vector3d position = args.<Vector3d>getOne("position").get();
                    World world = args.<WorldProperties>getOne("world")
                            .map(props -> Sponge.getServer().getWorld(props.getUniqueId()).get())
                            .orElseGet(((Locatable) src)::getWorld);

                    ParticleEffect.Builder builder = ParticleEffect.builder();
                    int count = args.<Integer>getOne("amount").orElse(1);

                    builder.type(particleType);
                    builder.count(count);
                    args.<Vector3d>getOne("offset").ifPresent(builder::offset);
                    args.<Vector3d>getOne("motion").ifPresent(builder::velocity);
                    args.<Vector3d>getOne("color").ifPresent(color ->
                            builder.option(ParticleOptions.COLOR, Color.of(color.toInt())));
                    args.<NotePitch>getOne("note").ifPresent(note ->
                            builder.option(ParticleOptions.NOTE, note));
                    args.<Double>getOne("scale").ifPresent(scale ->
                            builder.option(ParticleOptions.SCALE, scale));
                    args.<BlockState>getOne("block").ifPresent(blockState ->
                            builder.option(ParticleOptions.BLOCK_STATE, blockState));
                    args.<ItemType>getOne("item").ifPresent(item ->
                            builder.option(ParticleOptions.ITEM_STACK_SNAPSHOT, ItemStack.of(item, 1).createSnapshot()));

                    world.spawnParticles(builder.build(), position);
                    src.sendMessage(t("Successfully spawned the particle %s", particleType.getName()));
                    return CommandResult.success();
                })
                .build(), "particleeffect");
    }
}
```

### CODE REVIEW

File: src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java [hunk 1]
```diff
@@ -27,87 +27,91 @@
...
-    public static final ParticleType CRIT = DummyObjectProvider.createFor(ParticleType.class, "CRIT");
+    public static final ParticleType CRITICAL_HIT = DummyObjectProvider.createFor(ParticleType.class, "CRITICAL_HIT");
+
+    public static final ParticleType DAMAGE_INDICATOR = DummyObjectProvider.createFor(ParticleType.class, "DAMAGE_INDICATOR");
 
-    public static final ParticleType CRIT_MAGIC = DummyObjectProvider.createFor(ParticleType.class, "CRIT_MAGIC");
+    public static final ParticleType DRAGON_BREATH = DummyObjectProvider.createFor(ParticleType.class, "DRAGON_BREATH");
 
-    public static final ParticleType DRIP_LAVA = DummyObjectProvider.createFor(ParticleType.class, "DRIP_LAVA");
+    public static final ParticleType DRIP_ = DummyObjectProvider.createFor(ParticleType.class, "DRIP_LAVA");
```
Comment (Feeliiix) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java]: I think something went wrong here.

File: src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     * This option will affect how all the particles are spread.
+     */
+    public static final ParticleOption<Vector3d> OFFSET = dummy("OFFSET");
+
+    /**
+     * This option will change the scale of a particle, the only
+     * vanilla {@link ParticleType} that this option will affect is
+     * {@link ParticleTypes#LARGE_EXPLOSION}.
+     */
+    public static final ParticleOption<NotePitch> SCALE = dummy("SCALE");
```
Comment (Cybermaxke) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java]: Copy/paste fail :p

File: src/main/java/org/spongepowered/api/effect/particle/ParticleEffect.java [hunk 3]
```diff
@@ -26,12 +26,16 @@
...
+import org.spongepowered.api.data.persistence.DataBuilder;
+
+import java.util.Map;
+import java.util.Optional;
 
 /**
  * Represents a particle effect that can be send to the Minecraft client.
  */
-public interface ParticleEffect {
+public interface ParticleEffect extends DataSerializable {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleEffect.java]: Why is this extending `DataSerializable`?

File: src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * This option will affect how all the particles are moving.
+     */
+    public static final ParticleOption<Vector3d> VELOCITY = dummy("VELOCITY");
+
+    // SORTFIELDS:OFF
+
+    @SuppressWarnings("unchecked")
+    private static <V> ParticleOption<V> dummy(String name) {
+        return DummyObjectProvider.createFor(ParticleOption.class, name);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java]: Inline this to the field initializers.

File: src/main/java/org/spongepowered/api/effect/particle/ParticleEffect.java (hunk 3 above)
Comment (Cybermaxke) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleEffect.java]: @gabizou You may want to make particle effects configurable in a plugin, that is why I made it serializable.

**Comment by gabizou** [comment]
@Cybermaxke any updates on this?

**Comment by Cybermaxke** [comment]
@gabizou I updated the pr to 1.10 and fixed the ids in the implementation. I think that it's now finished.

**Comment by bloodmc** [comment]
@Cybermaxke can you rebase?

File: src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java [hunk 5]
```diff
@@ -43,6 +44,16 @@
...
+    /**
+     * This particle type will play the effect that will occur when
+     * a {@link Player} uses bone meal on a plant to boost the growth.
+     * <p>
+     * This type has limited {@link ParticleOption}s, only
+     * {@link ParticleOptions#BLOCK_STATE} and
+     * {@link ParticleOptions#ITEM_STACK_SNAPSHOT} are supported.
+     */
+    public static final ParticleType BREAK_BLOCK = DummyObjectProvider.createFor(ParticleType.class, "BREAK_BLOCK");
+
```
Comment (Maxqia) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java]: Erm...

**Comment by Zidane** [comment]
Need this updated to address the issues brought forward.

File: src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java (hunk 4 above)
Comment (Zidane) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java]: @Cybermaxke 

Can you do what @gabizou  said?

File: src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java [hunk 6]
```diff
@@ -24,90 +24,142 @@
...
-    public static final ParticleType.Resizable EXPLOSION_LARGE = DummyObjectProvider.createFor(ParticleType.Resizable.class, "EXPLOSION_LARGE");
+    public static final ParticleType EXPLOSION = DummyObjectProvider.createFor(ParticleType.class, "EXPLOSION");
 
-    public static final ParticleType EXPLOSION_NORMAL = DummyObjectProvider.createFor(ParticleType.class, "EXPLOSION_NORMAL");
+    public static final ParticleType FALLING_DUST = DummyObjectProvider.createFor(ParticleType.class, "FALLING_DUST");
+
+    /**
+     * This particle type will play the effect that will occurs when
+     * a {@link Player} uses bone meal on a plant to boost the growth.
+     * <p>
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java]: @gabizou 

Is this javadoc correct?

File: src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java (hunk 4 above)
Comment (Cybermaxke) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java]: @Zidane I used the dummy method too avoid that you have to add `@SuppressWarnings("unchecked")` to every field, or I could just suppress it for the whole class.

Comment (Zidane) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java]: Suppress it for the class.

File: src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java [hunk 7]
```diff
@@ -24,90 +24,148 @@
  */
 package org.spongepowered.api.effect.particle;
 
+import org.spongepowered.api.entity.living.player.Player;
 import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;
 
 /**
- * The particles that can be sent on a vanilla Minecraft client.
+ * An enumeration of all possible {@link ParticleType}s in vanilla minecraft.
```
Comment (liach) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java]: Should `minecraft` be in uppercase?

File: src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java [hunk 8]
```diff
@@ -95,6 +95,12 @@
      */
     public static final ParticleType FIRE_SMOKE = DummyObjectProvider.createFor(ParticleType.class, "FIRE_SMOKE");
 
+    /**
+     * This particle type will play a fireworks effect.
+     * <p>
+     * This type has limited {@link ParticleOption}s, only
+     * {@link ParticleOptions#FIREWORK_EFFECTS} is supported.
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java]: Need a closing `</p>` at the end of the paragraph.

File: src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java [hunk 9]
```diff
@@ -122,6 +126,10 @@
      *   <li>{@link ParticleTypes#LARGE_EXPLOSION}</li>
      *   <li>{@link ParticleTypes#SWEEP_ATTACK}</li>
      * <ul/>
+     * <p>
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java]: `<p>...</p>` (no break after `<p>` or before `</p>`)

File: src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java [hunk 10]
```diff
@@ -150,10 +158,17 @@
...
      * <ul>
+     *   <li>{@link ParticleTypes#BREAK_BLOCK}</li>
+     *   <li>{@link ParticleTypes#ENDER_TELEPORT}</li>
+     *   <li>{@link ParticleTypes#FIRE_SMOKE}</li>
+     *   <li>{@link ParticleTypes#FIREWORKS}</li>
      *   <li>{@link ParticleTypes#MOBSPAWNER_FLAMES}</li>
      *   <li>{@link ParticleTypes#SPLASH_POTION}</li>
-     *   <li>{@link ParticleTypes#BREAK_BLOCK}</li>
      * <ul/>
+     * <p>
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java]: `<p>...</p>` (no break after `<p>` or before `</p>`)

File: src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java [hunk 11]
```diff
@@ -51,6 +51,7 @@
      * This type has limited {@link ParticleOption}s, only
      * {@link ParticleOptions#BLOCK_STATE} and
      * {@link ParticleOptions#ITEM_STACK_SNAPSHOT} are supported.
+     * </p>
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/effect/particle/ParticleTypes.java]: `<p>...</p>` (no break after `<p>` or before `</p>`)

**Pull request closed without merging**